src/Compilers/Core/Portable/Text/SourceText.cs

```diff
@@ -1278,5 +1293,130 @@ internal static int GetMaxCharCountOrThrowIfHuge(Encoding encoding, Stream strea
+                    {
+                        Vector128.StoreUnsafe(TReverser.Reverse(Vector128.LoadUnsafe(ref sourceRef, (uint)i)), ref destRef, (uint)i);
+                        i += Vector128<T>.Count;
+                    }
+                }
+
+                while (i < source.Length)
+                {
+                    Unsafe.Add(ref destRef, i) = TReverser.Reverse(Unsafe.Add(ref sourceRef, i));
+                    i++;
+                }
+            }
+            else
+            {
+                // There's overlap between the source and the destination, and the source starts before the destination.
+                // That means if we were to iterate from beginning to end, reading from the source and writing to the
+                // destination, we'd overwrite source elements not yet read.  To avoid that, we iterate from end to beginning.
+
+                int i = source.Length;
+
+                if (Vector256.IsHardwareAccelerated)
+                {
+                    while (i >= Vector256<T>.Count)
+                    {
+                        i -= Vector256<T>.Count;
+                        Vector256.StoreUnsafe(TReverser.Reverse(Vector256.LoadUnsafe(ref sourceRef, (uint)i)), ref destRef, (uint)i);
+                    }
+                }
+
+                if (Vector128.IsHardwareAccelerated)
+                {
+                    while (i >= Vector128<T>.Count)
+                    {
+                        i -= Vector128<T>.Count;
+                        Vector128.StoreUnsafe(TReverser.Reverse(Vector128.LoadUnsafe(ref sourceRef, (uint)i)), ref destRef, (uint)i);
+                    }
+                }
+
+                while (i > 0)
+                {
+                    i--;
+                    Unsafe.Add(ref destRef, i) = TReverser.Reverse(Unsafe.Add(ref sourceRef, i));
+                }
+            }
+        }
+
+        private interface IEndiannessReverser<T> where T : struct
+        {
+            static abstract T Reverse(T value);
+            static abstract Vector128<T> Reverse(Vector128<T> vector);
+            static abstract Vector256<T> Reverse(Vector256<T> vector);
+        }
+
+        private readonly struct Int16EndiannessReverser : IEndiannessReverser<short>
+        {
+            public static short Reverse(short value) =>
+               BinaryPrimitives.ReverseEndianness(value);
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector128<short> Reverse(Vector128<short> vector) =>
+                Vector128.ShiftLeft(vector, 8) | Vector128.ShiftRightLogical(vector, 8);
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector256<short> Reverse(Vector256<short> vector) =>
+                Vector256.ShiftLeft(vector, 8) | Vector256.ShiftRightLogical(vector, 8);
+        }
+#else
+        private static void ReverseEndianness(ReadOnlySpan<short> source, Span<short> destination)
```

Comment:
if we rip the vectorized version out, this will take over.  won't be as fast, but is trivial to maintain and understand.